hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -1057,6 +1083,74 @@ private void validateRecordIndex(HoodieSparkEngineContext sparkEngineContext,
+    secondaryKeys.persist(StorageLevel.MEMORY_AND_DISK());
+    long numSecondaryKeys = secondaryKeys.count();
+    int numPartitions = (int) Math.max(1, numSecondaryKeys / 100);
+    secondaryKeys = secondaryKeys.sortBy(x -> x, true, numPartitions);
+    for (int i = 0; i < numPartitions; i++) {
+      List<String> secKeys = secondaryKeys.collectPartitions(new int[] {i})[0];
+      Map<String, List<String>> mdtSecondaryKeyToRecordKeys = ((HoodieBackedTableMetadata) metadataContext.tableMetadata)
+          .getSecondaryIndexRecords(secKeys, indexDefinition.getIndexName())
+          .entrySet().stream()
+          .collect(Collectors.toMap(Map.Entry::getKey,
+              e -> e.getValue().stream().map(rec -> rec.getData().getRecordKeyFromSecondaryIndex()).collect(Collectors.toList())));
+      Map<String, List<String>> fsSecondaryKeyToRecordKeys = getFSSecondaryKeyToRecordKeys(engineContext, basePath, latestCompletedCommit, indexDefinition.getSourceFields().get(0), secKeys);
+      if (!fsSecondaryKeyToRecordKeys.equals(mdtSecondaryKeyToRecordKeys)) {
+        throw new HoodieValidationException(String.format("Secondary Index does not match : \nMDT secondary index: %s \nFS secondary index: %s",
+            StringUtils.join(mdtSecondaryKeyToRecordKeys), StringUtils.join(fsSecondaryKeyToRecordKeys)));
+      }
+    }
+    secondaryKeys.unpersist();
+  }
+
+  Map<String, List<String>> getFSSecondaryKeyToRecordKeys(HoodieSparkEngineContext sparkEngineContext, String basePath, String latestCompletedCommit,
+                                                          String secondaryField, List<String> secKeys) {
+    List<Tuple2<String, String>> recordAndSecondaryKeys = sparkEngineContext.getSqlContext().read().format("hudi")
+        .option(DataSourceReadOptions.TIME_TRAVEL_AS_OF_INSTANT().key(), latestCompletedCommit)
+        .load(basePath)
+        .filter(functions.col(secondaryField).isin(secKeys.toArray()))
+        .select(RECORD_KEY_METADATA_FIELD, secondaryField)
+        .javaRDD()
+        .map(row -> new Tuple2<>(row.getAs(RECORD_KEY_METADATA_FIELD).toString(), row.getAs(secondaryField).toString()))
+        .collect();
+    Map<String, List<String>> secondaryKeyToRecordKeys = new HashMap<>();
+    for (Tuple2<String, String> recordAndSecondaryKey : recordAndSecondaryKeys) {
+      secondaryKeyToRecordKeys.compute(recordAndSecondaryKey._2, (k, v) -> {
+        List<String> recKeys = v != null ? v : new ArrayList<>();
+        recKeys.add(recordAndSecondaryKey._1);
+        return recKeys;
+      });
+    }
+    return secondaryKeyToRecordKeys;
+  }
+
+  private JavaRDD<String> readSecondaryKeys(HoodieSparkEngineContext engineContext, List<String> sourceFields, String basePath, String latestCompletedCommit) {
+    return engineContext.getSqlContext().read().format("hudi")
+        .option(DataSourceReadOptions.TIME_TRAVEL_AS_OF_INSTANT().key(), latestCompletedCommit)
```

Comment:
same here - why do we need to provide as of instant option?

Reply:
This is how the record indexes are also validated. I think it is good to have.